### PR TITLE
[quant][pt2e] Add some operaor support in QNNPackQuantizer

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99384
* __->__ #99381

Summary:
This PR adds support for flatten and also a variant of linear pattern to support quantizing an internal model

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45071007](https://our.internmc.facebook.com/intern/diff/D45071007)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45071007).

File: torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py [hunk 1]
```diff
@@ -255,13 +258,15 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
             # supported op type.
             self._annotate_conv2d_relu(node, global_config)
             self._annotate_conv2d(node, global_config)
-            self._annotate_linear(node, global_config)
+            self._annotate_linear3d(node, global_config)
+            self._annotate_linear2d(node, global_config)
```
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:262]: This is exactly what I dont like, but anywho. We need to close on module based stuff soon

Anyways we should put all of that innside annotate_linear.

File: torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py [hunk 2]
```diff
@@ -255,13 +258,15 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
...
             self._annotate_maxpool2d(node, global_config)
             self._annotate_add_relu(node, global_config)
             self._annotate_add(node, global_config)
             self._annotate_hardtanh(node, global_config)
             self._annotate_mean(node, global_config)
             self._annotate_adaptive_avg_pool2d(node, global_config)
+            self._annotate_flatten(node, global_config)
```
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:269]: I thought we were gonna propagate this via separate util?

File: torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:262]: yeah sure

File: torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py (hunk 2 above)
Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:269]: that's in a separate PR

**Comment by kimishpatel** [comment]
I am a bit worried here. Maybe take a step back to figure out the pattern matching stuff rather than doing it this way. Do you wanna update the stack so taht next diff can land? I am gonna try to use just subgraph pattern matching for linear till we figure out better solution.

**Comment by jerryzh168** [comment]
> 

why not land this to unblock lowering first and then figure out linear pattern matching stuff?

**Comment by kimishpatel** [comment]
> seems like too early to come up with an API

This is less about API and more of implementation details. I just dont want to keep adding tech debt.

**Comment by jerryzh168** [comment]
> > > 
> > 
> > 
> > why not land this to unblock lowering first and then figure out linear pattern matching stuff?
> > also TBH I feel optimizing for linear sounds like an early optimization to me, the scenario I'm thinking about that would need this immediately would be, e.g. there are 5-10 patterns like linear that we need to match and it becomes unmanageable to keep doing it, for now seems like too early to come up with an API, also we probably don't have enough data point yet I think?
> 
> Also this specific PR is addressing variants of linear and it is just resulting in hardcoding things that we are seeing for the models we want to support. I want to take this opportunity to do it in a robust way so at least for linear I want to be able to use the PR I put up, because only pattern matching option I want to consider is a) module partition or 2) subgraph matcher.

yeah this sounds good, we can use the pattern matching to support linear instead

**Comment by jerryzh168** [comment]
we'll use https://github.com/pytorch/pytorch/pull/99580 instead

**Pull request closed without merging**